README.md
# react-gallery-carousel

[![npm version][npm-badge]][npm-url]
[![npm downloads][downloads-badge]][npm-url]
[![npm bundle size][size-badge]][npm-url]
[![prettier][prettier-badge]][prettier-url]
[![license][license-badge]][license-url]
<a href="https://www.buymeacoffee.com/yifanai" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/default-orange.png" alt="Buy Me A Coffee" height="41" width="174"></a>

[npm-badge]: https://img.shields.io/npm/v/react-gallery-carousel.svg
[npm-url]: https://www.npmjs.com/package/react-gallery-carousel
[downloads-badge]: https://img.shields.io/npm/dm/react-gallery-carousel.svg
[size-badge]: https://badgen.net/bundlephobia/minzip/react-gallery-carousel
[prettier-badge]: https://img.shields.io/badge/code_style-prettier-ff69b4.svg
[prettier-url]: https://github.com/prettier/prettier
[license-badge]: https://img.shields.io/badge/license-MIT-blue.svg
[license-url]: https://github.com/yifaneye/react-gallery-carousel/blob/master/LICENSE

Mobile-friendly Carousel with batteries included (supporting touch, mouse emulation, lazy loading, thumbnails, fullscreen, RTL, keyboard navigation and customisations).

_The first version was published on 31st March 2021._

![Carousel controlled by finger](https://yifanai.s3-ap-southeast-2.amazonaws.com/rgc/demo_touch.gif)

Try **[Live Demo](https://yifanai.com/rgc)**

## Background

### What problems do other Carousels have?

I have **used and carefully analyzed a lot of carousel/slider components**. I summarized that **their issues** are:
1. Some of them do not move the slide as the user swipes on the slide.
2. Most of them do not support mouse dragging to move to the previous or the next slide. With the ones those support mouse emulation, some of them do not properly handle the case where the mouse leaves the carousel, which allow the user to continuously control the carousel.
3. Most of them do not support keyboard navigation (i.e. left, right and tab key).
4. Most of them cannot be maximized to fullscreen/modal/lightbox. With fullscreen, there is the issue of browser compatibility, i.e. iOS Safari on iPhone does not support the fullscreen API.
5. Most of them do not have an easy solution for building thumbnails. With the ones those have thumbnails, most of the thumbnails cannot be freely scrolled which lead to poor user experience. In addition, most of the thumbnails cannot be lazy loaded.
6. Most of them cannot lazy load (and preload) images. With the ones those can lazy load, most of them have transition that traverses the intermediate images when the user goes to a distant slide, which defeat the purpose of lazy loading.
7. Some of them cannot autoplay. With the ones those can autoplay, they cannot auto pause. For example, when the user hits another tab or goes to another app, the autoplay on those carousels do not pause.
8. Most of them do not respect the reduced motion settings by the user.
9. Most of them disregard the velocity of the swipe and just set a constant transition duration.
10. Some of their carousels will be in different sizes when the images/slides inside are in different sizes. Some of their transitions are bumpy when their images/slides are in different sizes.
11. Most of them do not support custom elements in a slide.
12. Most of them cannot be set to display in Right-to-Left (RTL).
13. Some of them disable pinching to zoom, while some others glitch when pinching with 2 fingers. Besides, when the window is zoomed in, most of them still detect for touch swiping to move to the previous, or the next slide, while the intention of most users in this scenario is panning to see other parts of the current slide.
14. Some of them will cause the slides to stuck its position on window resize or on mobile device orientation change, until another user interaction.
15. Some of them can only have predetermined images (i.e. before the carousel component mounts).
16. Most of them do not provide a solution for fallback/placeholder image (for when an image is not available).
17. Some of them get zoomed in when the user double taps on the control, while the intention of most users in this scenario is to quickly go to the next after the next slide.
18. Some of them remove the left or right button to indicate that there are no more slides in that direction. However, user is likely to click that spot where the button used to be, which causes undesired behaviours e.g. clicking on a link or button which is also at that spot.
19. Some of them use the method of cloning the first, and the last slide to achieve looping (or infinite mode). I think that method is not great semantically.
20. Some of them cannot distinguish a vertical swipe from a horizontal swipe, so that a not exactly vertical swipe moves the slides slightly horizontally; and a not exactly horizontal swipe moves the (document) page slightly vertically.

### What did I set to achieve?

I wanted to write my own **detail-oriented** and **exquisite** carousel component that is easy to use yet solves/supports all these things above under the hood. 🤓

I wanted to take my understanding of JavaScript events, DOM manipulation, browser APIs, cross-browser compatibility and performance debugging to the next level. 🤓

I wanted to master React functional components, hooks, custom hooks and reconciliation. 🤓

I wanted to learn more, place more care and attention to accessibility. I want to give focus outlines to the right users, support keyboard navigation, support screen reader, and follow [W3C accessible carousel tutorials](https://www.w3.org/WAI/tutorials/carousels/).

My carousel should support: touch, mouse emulation, keyboard navigation, modal (lightbox), thumbnails, autoplay (and auto pause), RTL (right to left for internationalization), image lazy loading (and preloading), responsive images, fallback image, reduced motion settings, instantaneous velocity detection, responsive design, images with any sizes, custom elements in a slide, pinch to zoom, customization and great accessibility. 😎

(e.g. To solve the last issue (#20) in the list above, my carousel should be able to detect a mostly vertical swipe and then fix the slides horizontally in the carousel. ✅
It should also be able to detect a mostly horizontal swipe and then fix the carousel vertically in the page. ✅)

## Demo

### Live demo
Try **[Live Demo](https://yifanai.com/rgc)**

### Carousel controlled by finger
![Carousel controlled by finger](https://yifanai.s3-ap-southeast-2.amazonaws.com/rgc/demo_touch.gif)

### Carousel controlled by cursor
![Carousel controlled by cursor](https://user-images.githubusercontent.com/49315663/118391709-1b490380-b679-11eb-9f56-44a6702419a0.gif)

### Lighthouse report
![Lighthouse Report on react-gallery-carousel](https://user-images.githubusercontent.com/49315663/118391896-f30dd480-b679-11eb-895b-e4cf3160303b.jpg)
![Lighthouse Report with scores on react-gallery-carousel](https://yifanai.s3-ap-southeast-2.amazonaws.com/rgc/lighthouse_report_v0.1.3.jpg)
This lighthouse report is conducted on https://yifaneye.github.io/react-gallery-carousel/ in an incognito window on Chrome version 89.0.4389.114 (Official Build) (x86_64) on MacBook Pro with macOS version 10.15.7 (19H2) on 12th of April 2021.

### Non-maximized carousels
![Non-maximized Carousel](https://yifanai.s3-ap-southeast-2.amazonaws.com/rgc/demo_non_maximized.jpg)

### Maximized carousel
![Maximized Carousel](https://yifanai.s3-ap-southeast-2.amazonaws.com/rgc/demo_maximized.jpg)

## Installation

```bash
npm install --save react-gallery-carousel
```
OR
```bash
yarn add react-gallery-carousel
```

## Usage

```jsx
import React from 'react';
import Carousel from 'react-gallery-carousel';
import 'react-gallery-carousel/dist/index.css';

const App = () => {
  const images = [9, 8, 7, 6, 5].map((number) => ({
    src: `https://placedog.net/${number}00/${number}00?id=${number}`
  }));

  return (
    <Carousel images={images} style={{ height: 500, width: 800 }} />
  );
};

export default App;

```

## Props

To customize the carousel, use the following props:

| Name                      | Type                    | Default                                | Description                                                                                                                                                                                                                                                                                                                                                                                                     |
|:--------------------------|:------------------------|:---------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| images                    | Array                   | undefined                              | Array of image(s) to be placed in the carousel. Each image object (e.g. [example object](#image-object-example)) in the array has a required attribute 'src'.                                                                                                                                                                                                                                                   |
| children                  | node or Array of nodes  | undefined                              | HTML element(s) to be placed into the carousel for user-managed slides. It (they) will be placed only if the 'images' prop is falsy.                                                                                                                                                                                                                                                                            |
| thumbnails                | Array of nodes          | undefined                              | HTML element(s) to be placed into the thumbnail view at the bottom. If not provided, the slides will be used. If provided, it must have the same length as the slides.                                                                                                                                                                                                                                          |
| index                     | Number                  | undefined                              | Initial (0-indexed) index of the slides of the carousel as a whole number starting from 0.                                                                                                                                                                                                                                                                                                                      |
| isRTL                     | Boolean                 | false                                  | If true, the slides of the carousel starts from the right (and also auto plays from the right to the left).                                                                                                                                                                                                                                                                                                     |
| isLoop                    | Boolean                 | true                                   | If true, the carousel form a loop (i.e. going left from the left-most slide lands at the right-most slide, and vice versa) from the ribbon of slides.                                                                                                                                                                                                                                                           |
| isMaximized               | Boolean                 | false                                  | If true, the carousel is maximized initially.                                                                                                                                                                                                                                                                                                                                                                   |
| shouldLazyLoad            | Boolean                 | true                                   | If true, images that are not yet in the viewport of the carousel will be lazy loaded, except the 2 adjacent images from either side of the carousel which will be preloaded.                                                                                                                                                                                                                                    |
| canAutoPlay               | Boolean                 | true                                   | If true, the carousel has auto play capability.                                                                                                                                                                                                                                                                                                                                                                 |
| isAutoPlaying             | Boolean                 | false                                  | If true, the carousel auto plays initially.                                                                                                                                                                                                                                                                                                                                                                     |
| autoPlayInterval          | Number                  | 5000                                   | Interval of the auto play (in milliseconds).                                                                                                                                                                                                                                                                                                                                                                    |
| hasTransition             | Boolean                 | true                                   | If false, the carousel does not have transition in moving between slides.                                                                                                                                                                                                                                                                                                                                       |
| swipeThreshold            | Number                  | 0.1                                    | Threshold swipe distance (in percentage of the width of the viewport of the carousel) to move to the previous or the next slide.                                                                                                                                                                                                                                                                                |
| swipeRollbackSpeed        | Number                  | 0.1                                    | Speed of the transition (in pixels per millisecond) in moving back to the current slide after a swipe smaller than swipeThreshold.                                                                                                                                                                                                                                                                              |
| transitionSpeed           | Number                  | 1                                      | Speed of the transition (in pixels per millisecond) in moving to the previous or the next slide on non-swipe updates on the carousel.                                                                                                                                                                                                                                                                           |
| transitionDurationLimit   | Number                  | 750                                    | Limit of transition duration (in milliseconds). The limit is used to flatten transition duration, where the maximum transition duration infinitely approaches this value.                                                                                                                                                                                                                                       |
| transitionDurationMin     | Number                  | 250                                    | Minimum transition duration (in milliseconds). Transition duration can be set to be a constant with transitionDurationMin === transitionDurationMax.                                                                                                                                                                                                                                                            |
| transitionDurationMax     | Number                  | undefined                              | Maximum transition duration (in milliseconds). It will have precedence over transitionDurationMin, if transitionDurationMin > transitionDurationMax.                                                                                                                                                                                                                                                            |
| widgetsHasShadow          | Boolean                 | false                                  | If true, the following widgets (with *) have shadows. (Note: if true, the transition will tend to drop frames when there are a large number of slides.)                                                                                                                                                                                                                                                         |
| hasLeftButton *           | Boolean or String       | 'centerLeft'                           | If false, the carousel does not show left button. Its position can be specified by one of [small widget positions](#small-widget-positions).                                                                                                                                                                                                                                                                    |
| hasRightButton *          | Boolean or String       | 'centerRight'                          | If false, the carousel does not show right button. Its position can be specified by one of [small widget positions](#small-widget-positions).                                                                                                                                                                                                                                                                   |
| hasMediaButton *          | Boolean or String       | 'topLeft'                              | If false, the carousel does not show media button (i.e. play/pause button). Its position can be specified by one of [small widget positions](#small-widget-positions). If the value of 'canAutoPlay' is falsy, media button will not be shown on the carousel.                                                                                                                                                  |
| hasSizeButton *           | Boolean or String       | 'topRight'                             | If false, the carousel does not show size button (i.e. maximize/minimize button). Its position can be specified by one of [small widget positions](#small-widget-positions).                                                                                                                                                                                                                                    |
| hasIndexBoard *           | Boolean or String       | 'topCenter'                            | If false, the carousel does not show index board (i.e. currentIndex / totalNumberOfSlides). Its position can be specified by one of [small widget positions](#small-widget-positions).                                                                                                                                                                                                                          |
| hasDotButtons *           | Boolean or String       | false                                  | If false, the carousel does not show dot buttons (i.e. array of dots indicating the current slide in relation to other slides). Its position can be specified by one of [large widget positions](#large-widget-positions).                                                                                                                                                                                      |
| hasCaptions *             | Boolean or String       | false                                  | If false, the carousel does not show caption for each image. Its position can be specified by one of [large widget positions](#large-widget-positions).                                                                                                                                                                                                                                                         |
| hasThumbnails             | Boolean                 | true                                   | If false, the carousel does not show thumbnails at the bottom.                                                                                                                                                                                                                                                                                                                                                  |
| hasLeftButtonAtMax *      | Boolean or String       | undefined                              | If false, the maximized carousel does not show left button. Its position can be specified by one of [small widget positions](#small-widget-positions). It overrides 'hasLeftButton' prop for the maximized carousel. If not specified, the value of 'hasLeftButton' will be used.                                                                                                                               |
| hasRightButtonAtMax *     | Boolean or String       | undefined                              | If false, the maximized carousel does not show right button. Its position can be specified by one of [small widget positions](#small-widget-positions). It overrides 'hasRightButton' prop for the maximized carousel. If not specified, the value of 'hasRightButton' will be used.                                                                                                                            |
| hasMediaButtonAtMax *     | Boolean or String       | undefined                              | If false, the maximized carousel does not show media button (i.e. play/pause button). Its position can be specified by one of [small widget positions](#small-widget-positions). It overrides 'hasMediaButton' prop for the maximized carousel. If not specified, the value of 'hasMediaButton' will be used. If the value of 'canAutoPlay' is falsy, media button will not be shown on the maximized carousel. |
| hasSizeButtonAtMax *      | Boolean or String       | undefined                              | If false, the maximized carousel does not show size button (i.e. maximize/minimize button). Its position can be specified by one of [small widget positions](#small-widget-positions). It overrides 'hasSizeButton' prop for the maximized carousel. If not specified, the value of 'hasSizeButton' will be used.                                                                                               |
| hasIndexBoardAtMax *      | Boolean or String       | undefined                              | If false, the maximized carousel does not show index board (i.e. currentIndex / totalNumberOfSlides). Its position can be specified by one of [small widget positions](#small-widget-positions). It overrides 'hasIndexBoard' prop for the maximized carousel. If not specified, the value of 'hasIndexBoard' will be used.                                                                                     |
| hasDotButtonsAtMax *      | Boolean or String       | undefined                              | If false, the maximized carousel does not show dot buttons (i.e. array of dots indicating the current slide in relation to other slides). Its position can be specified by one of [large widget positions](#large-widget-positions). It overrides 'hasDotButtons' prop for the maximized carousel. If not specified, the value of 'hasDotButtons' will be used.                                                 |
| hasCaptionsAtMax *        | Boolean or String       | undefined                              | If false, the maximized carousel does not show caption for each image. Its position can be specified by one of [large widget positions](#large-widget-positions). It overrides 'hasCaptions' prop for the maximized carousel. If not specified, the value of 'hasCaptions' will be used.                                                                                                                        |
| hasThumbnailsAtMax        | Boolean                 | undefined                              | If false, the maximized carousel does not show thumbnails at the bottom. It overrides 'hasThumbnails' for the maximized carousel. If not specified, the value of 'hasThumbnails' will be used.                                                                                                                                                                                                                  |

CHANGELOG.md
# ```react-gallery-carousel```

[![NPM](https://img.shields.io/npm/v/react-gallery-carousel.svg)](https://www.npmjs.com/package/react-gallery-carousel)

## v0.4.0 (2023-01-28)

### Enhancements

- Add support for custom elements.

- Improve format in the README file.

## v0.3.0 (2022-10-15)

### Bug Fixes

- Fix issue related to maximizing/minimizing not working when shouldMaximizeOnClick/shouldMinimizeOnClick is on, for Toggle Device Toolbar (simulated mobile devices).
  ([Issue #74](https://github.com/yifaneye/react-gallery-carousel/issues/74))

- Fix issue related to `npm install` not working for `/example`.
  ([Issue #72](https://github.com/yifaneye/react-gallery-carousel/issues/72))

### Enhancements

- Update package dependencies. ([Issue #71](https://github.com/yifaneye/react-gallery-carousel/issues/71))

## v0.2.10 (2022-08-03)

### Enhancements

- Add react 18 to peer dependencies.

## v0.2.9 (2022-01-30)

### Bug Fixes

- Third attempt to fix swiping stuck after zoom issue.
  ([Issue #44](https://github.com/yifaneye/react-gallery-carousel/issues/44))

## v0.2.8 (2021-12-23)

### Bug Fixes

- Second attempt to fix swiping stuck after zoom issue.
  ([Issue #44](https://github.com/yifaneye/react-gallery-carousel/issues/44))

### Enhancements

- Simplify Gatsby example

## v0.2.7 (2021-12-18)

### Enhancements

- Fix "invalid or unexpected token" issue in Next.js.
  ([Issue #51](https://github.com/yifaneye/react-gallery-carousel/issues/51))

- Remove request to "object Object"
  ([Issue #52](https://github.com/yifaneye/react-gallery-carousel/issues/52))

## v0.2.6 (2021-12-06)

### Enhancements

- Fix swiping stuck after zoom issue.
  ([Issue #44](https://github.com/yifaneye/react-gallery-carousel/issues/44))

- Remove title for icons.
  ([Issue #60](https://github.com/yifaneye/react-gallery-carousel/issues/60))

## v0.2.5 (2021-10-30)

### Enhancements

- Added **support for custom children for the thumbnails**.
  ([Issue #40](https://github.com/yifaneye/react-gallery-carousel/issues/40))

- Enhanced docs.

## v0.2.4 (2021-08-22)

### Enhancements

- Enhanced docs.
  ([Issue #43](https://github.com/yifaneye/react-gallery-carousel/issues/43))

- Enhanced code and examples.
- Set up momentum-based mouse dragging on the thumbnails.

## v0.2.3 (2021-07-12)

### Enhancements

- Made the code robuster.
  ([Issue #31](https://github.com/yifaneye/react-gallery-carousel/issues/31))

- Enhanced docs.
  ([Issue #35](https://github.com/yifaneye/react-gallery-carousel/issues/35))

- Added **support for server-side rendering (SSR)**.
  ([Issue #36](https://github.com/yifaneye/react-gallery-carousel/issues/36))

- Enhanced examples.
  ([Issue #45](https://github.com/yifaneye/react-gallery-carousel/issues/45))

## v0.2.2 (2021-05-16)

### Enhancements

- Enhanced the documentation regarding user-managed slides.
  ([Issue #28](https://github.com/yifaneye/react-gallery-carousel/issues/28))

- Updated the [demo site](https://yifanai.com/rgc).

## v0.2.1 (2021-05-08)

### Bug Fixes

- Fixed issue where caption for an image always has vertical and horizontal scroll bars even they are not required.
  ([Issue #26](https://github.com/yifaneye/react-gallery-carousel/issues/26))

### Enhancements

- Enhanced content on the [demo site](https://yifanai.com/rgc).

- Enhanced examples and content in the [documentation](https://yifanai.com/rgc).
  ([Issue #28](https://github.com/yifaneye/react-gallery-carousel/issues/28))

## v0.2.0 (2021-05-02)

### Bug Fixes

- Fixed a bug where left and right keys do not work when a widget is focused by listening keydown events on the entire carousel.
  ([Issue #15](https://github.com/yifaneye/react-gallery-carousel/issues/15))

- Fixed a bug where swiping on a slide does not freeze autoplay countdown.
  ([Issue #18](https://github.com/yifaneye/react-gallery-carousel/issues/18))

### Enhancements

- Allowed the size of the thumbnails to be customized.
  ([Issue #11](https://github.com/yifaneye/react-gallery-carousel/issues/11))

- Added handlers and listeners to the carousel.
  ([Issue #12](https://github.com/yifaneye/react-gallery-carousel/issues/12))

- Enabled smooth transition on the thumbnails.
  ([Issue #13](https://github.com/yifaneye/react-gallery-carousel/issues/13))

- Added aria-live region for accessibility.

## v0.1.4 (2021-04-18)

### Bug Fixes

- Fixed a minor bug where required props is undefined by defining default props.
  ([Issue #7](https://github.com/yifaneye/react-gallery-carousel/issues/7))

### Enhancements

- Enhanced content and styles on the [demo site](https://yifanai.com/rgc).

## v0.1.3 (2021-04-12)

### Bug Fixes

- Fixed a bug where touch swiping did not work on the slides when the images are dynamically set in the useEffect() hook by developer users.
  ([Issue #5](https://github.com/yifaneye/react-gallery-carousel/issues/5))

### Enhancements

- Enhanced content, styles, responsiveness and performance (from 98% to 100%) on the [demo site](https://yifanai.com/rgc).

## v0.1.2 (2021-04-11)

### Bug Fixes

- Fixed a bug where the last slide is not displayed starting from the second loop when there are exactly 2 slides.
  ([Issue #2](https://github.com/yifaneye/react-gallery-carousel/issues/2))

### Enhancements

- Removed the use of translate3d to make the performance consistent across browsers and platforms.
- Reduced the file size of the fallback image from 7.59 kB to 3.35 kB.
- Made the text displayed in the index board robuster.

## v0.1.1 (2021-04-10)

### Bug Fixes

- Fixed a bug where image URLs those are not known ahead of time cannot be placed into the carousel.
  ([Issue #1](https://github.com/yifaneye/react-gallery-carousel/issues/1))

### Enhancements

- Enabled hitting Enter (Return) key on the focused thumbnail to display that image as the current image.
- Enabled selecting text in the index board using cursor or finger.
- Removed the effect of hover on thumbnails on touch devices.

## v0.1.0 (2021-03-31)

### Initial work


src/Button/Button.js
import React from 'react';
import { Button } from 'react-responsive-button';
import 'react-responsive-button/dist/index.css';
import styles from './Button.module.css';
import PropTypes from 'prop-types';

export const ArrowButton = (props) => {
  const className = `${styles.buttonWrapper} ${styles[props.direction]}`;
  const icons = {
    left: {
      label: 'Previous Slide',
      image: `url("data:image/svg+xml,%3Csvg height='40' preserveAspectRatio='xMidYMid meet' viewBox='0 0 20 40' width='20' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='m20 0v40l-20-20z' fill='%23eee' fill-opacity='.5'/%3E%3C/svg%3E")`
    },
    right: {
      label: 'Next Slide',
      image: `url("data:image/svg+xml,%3Csvg height='40' preserveAspectRatio='xMidYMid meet' viewBox='0 0 20 40' width='20' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='m0 40v-40l20 20z' fill='%23eee' fill-opacity='.5'/%3E%3C/svg%3E")`
    }
  };
  const { label, image } = icons[props.direction];

  return (
    <div className={className}>
      <Button
        disabled={props.disabled}
        shape='rectangle'
        aria-label={label}
        aria-disabled={false}
        style={{
          width: 20,
          height: 40,
          backgroundColor: 'transparent',
          backgroundImage: image,
          backgroundSize: '20px 40px',
          backgroundPosition: 'center',
          backgroundRepeat: 'no-repeat'
        }}
        onClick={props.clickCallback}
      />
    </div>
  );
};

ArrowButton.propTypes = {
  direction: PropTypes.oneOf(['left', 'right']).isRequired,
  disabled: PropTypes.bool,
  clickCallback: PropTypes.func
};

export const MediaButton = (props) => {
  const icons = {
    play: {
      label: 'Play Auto Play',
      image: `url("data:image/svg+xml,%3Csvg height='30' preserveAspectRatio='xMidYMid meet' viewBox='0 0 30 30' width='30' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='m15 22.5-15 7.5v-15-15l15 7.5 15 7.5z' fill='%23eee' fill-opacity='.5'/%3E%3C/svg%3E")`
    },
    pause: {
      label: 'Pause Auto Play',
      image: `url("data:image/svg+xml,%3Csvg height='30' preserveAspectRatio='xMidYMid meet' viewBox='0 0 30 30' width='30' xmlns='http://www.w3.org/2000/svg'%3E%3Cg fill='%23eee' fill-opacity='.5'%3E%3Cpath d='m0 0h10v30h-10z'/%3E%3Cpath d='m20 0h10v30h-10z'/%3E%3C/g%3E%3C/svg%3E")`
    }
  };
  const { label, image } = icons[props.isPlaying ? 'pause' : 'play'];

  const className = `${styles.buttonWrapper} ${styles.media}`;
  return (
    <div className={className}>
      <Button
        disabled={props.disabled}
        shape='rectangle'
        aria-label={label}
        aria-disabled={false}
        style={{
          width: 30,
          height: 30,
          backgroundColor: 'transparent',
          backgroundImage: image,
          backgroundSize: '30px 30px',
          backgroundPosition: 'center',
          backgroundRepeat: 'no-repeat'
        }}
        onClick={props.clickCallback}
      />
    </div>
  );
};

MediaButton.propTypes = {
  isPlaying: PropTypes.bool.isRequired,
  disabled: PropTypes.bool,
  clickCallback: PropTypes.func
};
